### ISSUE TITLE
Support for history.pushState()

### ISSUE THREAD

**Issue by NicBright** [angular/zone.js#1161]
I just found out that history.pushState() is not patched.

I would have expected it to be patched because it's most natural to expect the Angular router to pick up changes to the url from outside of Angular just as well. 

For tracking purposes:
In this issue from 2014, btford already outlined History API to be supported by Zone.js: https://github.com/angular/zone.js/issues/9

**Comment by JiaLiPassion** [comment]
@NicBright, not sure what need to do to patch `pushState`, it is just a `function`, without any `callback` or `async operations`.

**Comment by NicBright** [comment]
@JiaLiPassion, yes it's "just a function". As far as I know there's no event triggered when pushState() is called. I can imagine a simple monkey patch like this:

```
const pushStateOriginal = history.pushState.bind(history);
history.pushState = (...args) => {
  const result = pushStateOriginal(...args);
  Zone.run(); // I don't know the Zone.js details, but that'd be the point where Zone would hook into to do its magic, like invoking the "Router (url-)change detection"
  return result;
}
```

**Comment by JiaLiPassion** [comment]
@NicBright, could you post a use case or some reproduce repo to describe that without the patch change detection will not run? Thanks!

**Comment by NicBright** [comment]
@JiaLiPassion my concrete use case is this:
In my Angular SPA I'm loading a 3rd party application (not necessarily an Angular SPA) within an iframe. Inside the 3rd party application there's a link that shall trigger navigation of the Angular SPA ("create a new business object starting with pre-defined parameters"). Until now, we used the hashbang navigation mode where everything worked well, because there was the "hashchange" event triggered when we did `window.top.location = '#/tmp/createObject/...'`. Unfortunately `window.top.location` completely ignores HTML5 location mode and will **always** trigger a reload of the whole SPA. The only way I could come up with is to change it to `window.top.history.pushState({}, "...", '/tmp/createObject/...')`. However, this will not immediately trigger the Router to catch up. Instead, you need to click in the Angular SPA for change detection to run and catch up. This is not acceptable, however. This forces us to continue using the `window.top.location=...` way, but again, this triggers a reload of the whole SPA (but at least it reloads instantly without further user interaction).

**Comment by JiaLiPassion** [comment]
I see, I will try to create a reproduce repo, thanks!

**Comment by JiaLiPassion** [comment]
I will check it this weekend.
